admission-webhook/integration_tests/integration_test.go

```diff
@@ -0,0 +1,331 @@
+	// this is the JSON representation of the cred spec from templates/credspec-2.yml
+	expectedCredSpec2 = `{"ActiveDirectoryConfig":{"GroupManagedServiceAccounts":[{"Name":"WebApplication2","Scope":"CONTOSO"},{"Name":"WebApplication2","Scope":"contoso.com"}]},"CmsPlugins":["ActiveDirectory"],"DomainJoinConfig":{"DnsName":"contoso.com","DnsTreeName":"contoso.com","Guid":"244818ae-87ca-4fcd-92ec-e79e5252348a","MachineAccountName":"WebApplication2","NetBiosName":"CONTOSO","Sid":"S-1-5-21-2126729477-2524275714-3294792973"}}`
+
+	tmpRoot      = "tmp"
+	ymlExtension = ".yml"
+)
+
+func TestHappyPathWithPodLevelAnnotation(t *testing.T) {
+	testName := "happy-path-with-pod-level-annotation"
+	credSpecTemplates := []string{"credspec-0"}
+	templates := []string{"credspecs-users-rbac-role", "service-account", "sa-rbac-binding", "simple-with-gmsa"}
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, credSpecTemplates, templates)
+	defer tearDownFunc()
+
+	pod := waitForPodToComeUp(t, testConfig.Namespace, "app="+testName)
+
+	assert.Equal(t, expectedCredSpec0, pod.Annotations["pod.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+}
+
+func TestHappyPathWithContainerLevelAnnotation(t *testing.T) {
+	testName := "happy-path-with-container-level-annotation"
+	credSpecTemplates := []string{"credspec-0"}
+	templates := []string{"credspecs-users-rbac-role", "service-account", "sa-rbac-binding", "simple-with-container-level-gmsa"}
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, credSpecTemplates, templates)
+	defer tearDownFunc()
+
+	pod := waitForPodToComeUp(t, testConfig.Namespace, "app="+testName)
+
+	assert.Equal(t, expectedCredSpec0, pod.Annotations["nginx.container.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+}
+
+func TestHappyPathWithSeveralContainers(t *testing.T) {
+	testName := "happy-path-with-several-containers"
+	credSpecTemplates := []string{"credspec-0", "credspec-1", "credspec-2"}
+	templates := []string{"credspecs-users-rbac-role", "service-account", "sa-rbac-binding", "several-containers-with-gmsa"}
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, credSpecTemplates, templates)
+	defer tearDownFunc()
+
+	pod := waitForPodToComeUp(t, testConfig.Namespace, "app="+testName)
+
+	assert.Equal(t, expectedCredSpec0, pod.Annotations["nginx0.container.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+	assert.Equal(t, expectedCredSpec1, pod.Annotations["pod.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+	assert.Equal(t, expectedCredSpec2, pod.Annotations["nginx2.container.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+}
+
+func TestHappyPathWithPreSetMatchingAnnotations(t *testing.T) {
```

Comment:
Is this different from `TestHappyPathWithPodLevelAnnotation` ?

Reply:
Yes, this one already defines its cred spec's contents in the annotations. The contents are actually in a different "order" (ie the keys in the JSON object) in the corresponding yml file; but for some reason they end up being the same after the pod's creation, which kinda baffles me. Was actually planning to spend a little time this morning getting to the bottom of this.

Reply:
(I'm dumb, I wasn't using the right yml...)